proof/refine/ARM/Tcb_R.thy

```diff
@@ -24,47 +24,262 @@ lemma activate_idle_thread_corres:
     (arch_activate_idle_thread t) (activateIdleThread t)"
   by (simp add: arch_activate_idle_thread_def activateIdleThread_def)
 
+lemma gts_st_tcb':
+  "\<lbrace>tcb_at' t\<rbrace> getThreadState t \<lbrace>\<lambda>rv. st_tcb_at' (\<lambda>st. st = rv) t\<rbrace>"
+  apply (rule hoare_vcg_precond_imp)
+   apply (rule hoare_post_imp[where Q="\<lambda>rv s. \<exists>rv'. rv = rv' \<and> st_tcb_at' (\<lambda>st. st = rv') t s"])
+    apply simp
+   apply (wp hoare_ex_wp)
+  apply (clarsimp simp add: pred_tcb_at'_def obj_at'_def)
+  done
+
+lemma activateIdle_invs:
+  "activateIdleThread thread \<lbrace>invs'\<rbrace>"
+  by (simp add: activateIdleThread_def)
+
+lemma invs'_live_sc'_ex_nonz_cap_to':
+  "ko_at' ko scp s \<Longrightarrow> invs' s \<Longrightarrow> live_sc' ko \<longrightarrow> ex_nonz_cap_to' scp s"
+  apply (clarsimp simp: invs'_def valid_state'_def if_live_then_nonz_cap'_def)
+  by (fastforce simp: obj_at'_real_def ko_wp_at'_def projectKO_sc)
+
+lemma invs'_ko_at_valid_sc:
```

Comment:
I'm much rather use `invs_valid_objs'` and then use the existing rule `sc_ko_at_valid_objs_valid_sc'`